### ISSUE TITLE
x/playground: do the "vet" check along with running a program

### ISSUE THREAD

**Issue by ysmolsky** [golang/go#24576]
"vet" is implemented as a separate button. Consider if we can remove that button and do the "vet" check whenever a user pushes the "run" button.

Comments from [CL 100776](https://golang.org/cl/100776):

> **Bryan Mills**
> I wonder whether this should even be a separate button. Could we Vet unconditionally whenever we Run?
> 
> We already have a bit of a visual distinction between “errors” and “output”, so it probably wouldn't be terribly confusing to put the errors before the output if we have both, and that would make the warnings that much more discoverable.
> 
> **Nathan Youngman**
> I think that's a good point.
> 
> Also, with the Go Playground supporting tests, and tests in Go 1.10 now running a subset of vet, it seems like this could just be automatic whether running tests or code. Though it may be important to clearly indicate what is an error from the compiler vs. an error from vet for people new to Go.
> 
> https://go-review.googlesource.com/c/playground/+/98155
> 
> **Andrew Bonventre**
> This would have ramifications for every call to /compile, which is an endpoint that anyone can use right now.
> 
> **Yury Smolsky**
> That would mean changing the spec of /compile endpoint. That endpoint is not supposed to return errors along with output. All clients assume that if there is an error, then output should not be considered. So this could break clients. IMHO, this feature requires new version of /compile along with new protocol.
> 
> **Nathan Youngman**
> There is more than one way to do it. An alternative is for the front-end to call /compile and /vet from the same button click. 

https://go-review.googlesource.com/c/playground/+/100776/5/edit.html#115

**Comment by opennota** [comment]
No way. Refusing to run perfectly correct programs isn't what one expects of a playground, for Zeus' sake.

`fmt.Println("Is it?\n") // vet: redundant newline`

**Comment by ysmolsky** [comment]
It is not supposed that "vet" errors will prevent from executing the program. Errors from vet must be displayed along with output from the program. However how to present it clearly, I don't know.

**Comment by rsc** [comment]
Back end protocol aside (that can be worked out either way), it does seem nice to just show the vet output alongside the normal output, instead of having yet another button. Unlike `go test`, if vet fails, the program should still run. (This is the playground.)

**Comment by nathany** [comment]
Now that the playground supports `go test` (#6511) is there any interplay to consider between `go test` running vet and running vet standalone? Or would the vet part of `go test` just be disabled/ignored in the playground in favour of the same vet as non-test code?

Regarding back end, what makes most sense to me at the moment is separate vet and compile endpoints that the UI can use from the same button click. What happens when the compile fails with errors though? Does vet only run after the compile succeeds and the output (begins) to display? (Keeping in mind time.Sleep works in the playground).

**Comment by ysmolsky** [comment]
To clarify `go test` function of Playground. Right now it does not run `vet` while testing. So we assume that we have to run `go vet` for the code containing tests too. And errors returned from `go vet` should not break an execution of a program even for tests. It differs from the behaviour of a real 'go test', but this will provide some consistency for the playground users.

Some examples of how it might look.

A1. vet errors after the output. User would have to wait until all output is replayed first.:

![screen shot 2018-04-14 at 10 45 02](https://user-images.githubusercontent.com/140245/38765899-a0ac534a-3fd2-11e8-9efa-678b8ad68cc3.png)

A2. A test:

![screen shot 2018-04-14 at 10 45 19](https://user-images.githubusercontent.com/140245/38765904-a5a0be22-3fd2-11e8-87fb-664f2cec0758.png)

B1. If we put go vet errors before the output and with no prefix:

![screen-shot-2018-04-14-at-10 45 02-b](https://user-images.githubusercontent.com/140245/38765907-b5d958da-3fd2-11e8-84ce-6a706efc4053.png)

B2. A test:

![screen-shot-2018-04-14-at-10 45 19-b](https://user-images.githubusercontent.com/140245/38765909-ba3d9ff8-3fd2-11e8-836a-555132a22498.png)

Any suggestion how to improve A variant?

**Comment by ysmolsky** [comment]
This one is my favourite. I have made the output of stderr to look like an error and put vet errors before the output. It is not indicated that first error is from `vet`. Should we?

![screen shot 2018-04-15 at 10 16 20](https://user-images.githubusercontent.com/140245/38775904-3983e146-4096-11e8-919f-7935f989fc2d.png)

Running an example without output compiles it but not executes producing this error:

![screen shot 2018-04-15 at 10 19 28](https://user-images.githubusercontent.com/140245/38775922-a35b1dbe-4096-11e8-810e-c800f3f2f7a0.png)

**Comment by bcmills** [comment]
All of these formats seem to rely on differences in color, which might not work so well for someone with red-green colorblindness, although I suppose that's also an existing issue.

To distinguish `vet` output from user code, what about putting the vet output at the end of the affected line instead of after the program?

That might look like:
![screenshot 2018-04-16 at 15 56 29](https://user-images.githubusercontent.com/5200974/38831952-f2942454-418e-11e8-9190-1ac5b8f5b089.png)

That way the warning isn't buried at the end of the output, but also doesn't bury the output. On the other hand, line-wrapping makes things a bit dicey for small windows, long code lines, and/or long error lines.

**Comment by jimmyfrasche** [comment]
Red/black can look a lot like black/black or black/nearly-black for various forms of color blindness and mostly for red-green color blindness which is the most prevalent form.

Another way to distinguish the vet output is to leave it where it is but draw a box around it labeled something like "vet output".

The red issue also applies to stdout vs stderr. Maybe adding unhighlightable {out, err} labels where line numbers would go in addition to the red? They could be made smallish since they'd essentially be icons. The lack of either on vet warnings would probably be enough to make it clear where things start and stop along with a little more spacing between the vet output and the program output.

**Comment by ysmolsky** [comment]
More screenshots for my proposed solution/design above: https://imgur.com/a/Jn3qx. 

Thanks for the feedback. I will have to meditate on it.

**Comment by ysmolsky** [comment]
I don't want to add graphics for this feature. I really like simple or minimal design as it is right now. Also I want vet errors to be on the top, that they urge a user to fix the code and get rid of them. They must be annoying in some sense.

Is current design friendly to color blind people? 

I propose this simple layout:
![screen shot 2018-04-20 at 23 39 02](https://user-images.githubusercontent.com/140245/39072960-db17f99a-44f4-11e8-8389-15eda984ad9e.png)

If go vet returned no errors, then first 3 lines are not displayed.

**Comment by jimmyfrasche** [comment]
The explicit delimiter works. It's not friendly, per se—but it's not antagonistic.

**Comment by ysmolsky** [comment]
One more observation:

```
package main

func main() {
	qwe
}
```
produces
```
prog.go:4:2: undefined: qwe

Program exited.
```

But in fact the program was not running at all. Error came from the go build. So the output could be:
```
prog.go:4:2: undefined: qwe

Go build exited.
```

**Comment by jimmyfrasche** [comment]
"Go build failed." might be clearer. It always exits.

**Comment by bcmills** [comment]
We should ensure that vet results show up in the Tour, too. That might save some folks a lot of time while they're learning.

**Comment by ysmolsky** [comment]
I agree. Tour is a different thing and does things differently. It uses `golang.org/compile`. Since we add new endpoint for vetting, we do not break Tour by any means. So that feature is still possible and should be tackled as a separate change.

I have updated https://go-review.googlesource.com/c/tools/+/107455 with the proposal I made above. I think this is the best design retaining simplicity I can think of. Examples are [here](https://imgur.com/a/kwuvExh).
